storage/tree_validation.go

```diff
@@ -64,6 +64,21 @@ func ValidateTreeForCreation(ctx context.Context, tree *trillian.Tree) error {
 	return validateMutableTreeFields(ctx, tree)
 }
 
+func validateTreeTypeUpdate(oldTree, newTree *trillian.Tree) error {
+	const prefix = "changing tree_type"
+	const wantState = trillian.TreeState_FROZEN
+	if oldState := oldTree.TreeState; oldState != wantState {
```

Comment:
Maybe the first two ifs could be combined? Then it might be clearer that the frozen state can't changed, with the following if that allows a preordered_log -> log type change.

Reply:
Done to some extent. I deliberately do 2 ifs for the TreeState checks to make error reporting detailed.